library/init/category/state.lean

```diff
@@ -1,124 +1,131 @@
 /-
 Copyright (c) 2016 Microsoft Corporation. All rights reserved.
 Released under Apache 2.0 license as described in the file LICENSE.
-Authors: Leonardo de Moura
+Authors: Leonardo de Moura, Sebastian Ullrich
+
+The state monad transformer.
 -/
 prelude
-import init.meta.interactive
-universes u v
-
-def state (σ α : Type u) : Type u :=
-σ → α × σ
+import init.category.alternative init.category.lift
+import init.category.id init.category.except
+universes u v w
 
-section
-variables {σ α β : Type u}
-
-@[inline] def state_return (a : α) : state σ α :=
-λ s, (a, s)
-
-@[inline] def state_bind (a : state σ α) (b : α → state σ β) : state σ β :=
-λ s, match (a s) with (a', s') := b a' s' end
-
-instance (σ : Type u) : monad (state σ) :=
-{pure := @state_return σ, bind := @state_bind σ,
- id_map := begin
-   intros, funext,
-   simp [state_bind], cases x s,
-   apply rfl
- end,
- pure_bind := by intros; apply rfl,
- bind_assoc := begin
-   intros, funext,
-   simp [state_bind], cases x s,
-   apply rfl
- end}
-end
+structure state_t (σ : Type u) (m : Type u → Type v) (α : Type u) : Type (max u v) :=
+(run' : σ → m (α × σ))
 
-namespace state
-@[inline] def read {σ : Type u} : state σ σ :=
-λ s, (s, s)
+attribute [pp_using_anonymous_constructor] state_t
 
-@[inline] def write {σ : Type} : σ → state σ unit :=
-λ s' s, ((), s')
+@[reducible] def state (σ α : Type u) : Type u := state_t σ id α
 
-@[inline] def write' {σ : Type u} : σ → state σ punit :=
-λ s' s, (punit.star, s')
-end state
+namespace state_t
+section
+  variables {σ : Type u} {m : Type u → Type v}
 
-def state_t (σ : Type u) (m : Type u → Type v) [monad m] (α : Type u) : Type (max u v) :=
-σ → m (α × σ)
+  -- swap `st` and `x`
```

Comment:
Why do you want to swap `st` and `x`?
The only motivation I found is this term here `(state_t.run s ∘ c)`.
With `run'`, you would have to write `(λ a, state_t.run' (c a) s)`.

Reply:
For the parser proofs, I found the goal state to be much more readable when the program comes last. But I defined my own `run` there anyway, so I removed this one now.